batch/batch/server/database.py

```diff
@@ -1,159 +1,133 @@
+                  'main_log_uri': 'VARCHAR(1000)',
+                  'output_log_uri': 'VARCHAR(1000)'}
+
+        keys = ['id']
+
+        await super().__init__(db, name, schema, keys)
+
+    async def update_record(self, id, **items):
+        await super(JobsTable, self).update_record({'id': id}, items)
+
+    async def get_records(self, ids, fields=None):
+        assert isinstance(ids, list)
+        return await super(JobsTable, self).get_record({'id': id}, fields)
+
+    async def get_record(self, id, fields=None):
+        records = await self.get_records({'id': id}, fields)
+        assert len(records) == 1
+        return records[0]
+
+    async def has_record(self, id):
+        return await super(JobsTable, self).has_record({'id': id})
+
+    async def delete_record(self, id):
+        await super(JobsTable, self).delete_record({'id': id})
+
+    async def get_incomplete_parents(self, id):
+        parent_ids = await self.get_record(id, ['parent_ids'])
+        parent_ids = json.loads(parent_ids['parent_ids'])
+        parent_records = await self.get_records(parent_ids)
+        incomplete_parents = [pr['id'] for pr in parent_records.result()
+                              if pr['state'] == 'Created']
+        return incomplete_parents
+
+
+class JobsParentsTable(Table):
+    async def __init__(self, db, name='jobs-parents'):
+        schema = {'id': 'BIGINT',
+                  'parent': 'BIGINT'}
+        keys = []
+
+        await super().__init__(db, name, schema, keys)
+
+    async def get_parents(self, id):
+        result = await super(JobsParentsTable, self).get_record({'id': id}, ['parent'])
+        return [record['parent'] for record in result]
+
+    async def get_children(self, id):
+        result = await super(JobsParentsTable, self).get_record({'parent': id}, ['id'])
+        return [record['id'] for record in result]
+
+
+class BatchTable(Table):
+    async def __init__(self, db, name='batch'):
+        schema = {'id': 'BIGINT NOT NULL AUTO_INCREMENT',
+                  'attributes': 'TEXT(65535)',
+                  'callback': 'TEXT(65535)',
+                  'ttl': 'INT',
+                  'is_open': 'BOOLEAN NOT NULL'
+                  }
+        keys = ['id']
+
+        await super().__init__(db, name, schema, keys)
+
+    async def update_record(self, id, **items):
+        await super(BatchTable, self).update_record({'id': id}, items)
+
+    async def get_records(self, ids, fields=None):
+        assert isinstance(ids, list)
+        return await super(BatchTable, self).get_record({'id': id}, fields)
+
+    async def get_record(self, id, fields=None):
+        records = await self.get_records({'id': id}, fields)
+        assert len(records) == 1
+        return records[0]
+
+    async def has_record(self, id):
+        return await super(BatchTable, self).has_record({'id': id})
+
+
+class BatchJobsTable(Table):
+    async def __init__(self, db, name='batch-jobs'):
+        schema = {'id': 'BIGINT',
```

Comment:
again, id is probably meant to be `NOT NULL AUTO_INCREMENT`?

Reply:
No. This is a relation from the id in `batch` to `jobs`.

Reply:
Ok, I would name this something other than 'id'. `table`_id